compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/PromisedValue.kt

```diff
@@ -61,103 +94,73 @@ abstract class BooleanValue(codegen: ExpressionCodegen) : PromisedValue(codegen,
-            val value = this@coerce.materialized
-            StackValue.coerce(value.type, type, mv)
-        }
-    }
+
+fun PromisedValue.materializedAt(target: Type, irTarget: IrType): MaterialValue {
+    materializeAt(target, irTarget)
+    return MaterialValue(codegen, target, irTarget)
 }
 
-fun PromisedValue.coerce(irTarget: IrType) =
-    coerce(typeMapper.mapType(irTarget), irTarget)
+fun PromisedValue.materialized(): MaterialValue =
+    materializedAt(type, irType)
 
-fun PromisedValue.coerceToBoxed(irTarget: IrType) =
-    coerce(typeMapper.boxType(irTarget), irTarget)
+fun PromisedValue.materializedAt(irTarget: IrType): MaterialValue =
+    materializedAt(typeMapper.mapType(irTarget), irTarget)
 
-// Same as above, but with a return type that allows conditional jumping.
-fun PromisedValue.coerceToBoolean() =
-    when (val coerced = coerce(Type.BOOLEAN_TYPE, codegen.context.irBuiltIns.booleanType)) {
-        is BooleanValue -> coerced
-        else -> object : BooleanValue(codegen) {
-            override fun jumpIfFalse(target: Label) = coerced.materialize().also { mv.ifeq(target) }
-            override fun jumpIfTrue(target: Label) = coerced.materialize().also { mv.ifne(target) }
-            override fun materialize() = coerced.materialize()
-        }
-    }
+fun PromisedValue.materializedAtBoxed(irTarget: IrType): MaterialValue =
+    materializedAt(typeMapper.boxType(irTarget), irTarget)
 
-// Non-materialized value of Unit type
-// This value is only materialized when calling coerce, which is why it is represented
-// as a MaterialValue even though there is nothing on the stack.
-val ExpressionCodegen.immaterialUnitValue: ImmaterialValue
-    get() = ImmaterialValue(this, Type.VOID_TYPE, context.irBuiltIns.unitType)
-
-// Non-materialized default value for the given type
-fun ExpressionCodegen.defaultValue(irType: IrType): PromisedValue =
-    object : PromisedValue(this, typeMapper.mapType(irType), irType) {
-        override fun materialize() {
-            StackValue.coerce(Type.VOID_TYPE, type, codegen.mv)
-        }
-    }
+fun PromisedValue.materialize() {
+    materializeAt(type, irType)
+}
 
-private fun IrType.isArrayOfKClass(): Boolean =
-    isArray() && this is IrSimpleType && arguments.singleOrNull().let { argument ->
-        argument is IrTypeProjection && argument.type.isKClass()
-    }
+fun PromisedValue.materializeAt(irTarget: IrType) {
+    materializeAt(typeMapper.mapType(irTarget), irTarget)
+}
+
+fun PromisedValue.materializeAtBoxed(irTarget: IrType) {
+    materializeAt(typeMapper.boxType(irTarget), irTarget)
+}
+
+val IrType.unboxed: IrType
+    get() = InlineClassAbi.getUnderlyingType(erasedUpperBound)
+
+// A Non-materialized value of Unit type that is only materialized through coercion.
+val ExpressionCodegen.unitValue: PromisedValue
+    get() = object : PromisedValue(this, Type.VOID_TYPE, context.irBuiltIns.unitType) {
+        override fun discard() {}
+    }
```

Comment:
I think it makes more sense to use `MaterialValue` here, since it can be considered a "materialized" value of `VOID_TYPE`, a type, any value of which takes zero space on the stack.
```suggestion
    get() = MaterialValue(this, Type.VOID_TYPE, context.irBuiltIns.unitType)
```